src/core/jsonschema/include/sourcemeta/core/jsonschema_vocabularies.h

```diff
@@ -0,0 +1,165 @@
+  uint32_t enabled_known = 0;
+
+  /// Bitflags for disabled known vocabularies (for explicit false values)
+  uint32_t disabled_known = 0;
+
+  /// Fallback storage for custom/unknown vocabularies
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4251)
+#endif
+  std::unordered_map<JSON::String, bool> custom;
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+
+  /// Default constructor
+  Vocabularies() = default;
+
+  /// Construct from initializer list (for backward compatibility)
+  Vocabularies(std::initializer_list<std::pair<JSON::String, bool>> init) {
+    for (const auto &entry : init) {
+      insert(entry);
+    }
+  }
+
+  /// Check if a vocabulary is enabled
+  /// @param uri The vocabulary URI to check
+  /// @return true if the vocabulary is enabled, false otherwise
+  [[nodiscard]] auto contains(std::string_view uri) const -> bool;
+
+  /// Insert a vocabulary with its enabled/disabled status
+  /// @param entry A pair of vocabulary URI and its boolean status
+  auto insert(const std::pair<JSON::String, bool> &entry) -> void;
+
+  /// Get vocabulary status by URI
+  /// @param uri The vocabulary URI to look up
+  /// @return std::optional<bool> containing the status if found, std::nullopt
+  /// otherwise
+  [[nodiscard]] auto find(std::string_view uri) const -> std::optional<bool>;
+
+  /// Merge another Vocabularies into this one.
+  /// Only adds vocabularies that don't already exist in this instance.
+  /// Matches std::unordered_map::merge semantics.
+  /// @param other The Vocabularies instance to merge from
+  auto merge(const Vocabularies &other) -> void;
+
+  /// Get all active (enabled or disabled) vocabularies as key-value pairs
+  /// @return A vector of pairs containing vocabulary URIs and their status
+  [[nodiscard]] auto all_vocabularies() const
+      -> std::vector<std::pair<std::string, bool>>;
+
+  /// Get the number of vocabularies (enabled + disabled + custom)
+  /// @return The total count of vocabularies
+  [[nodiscard]] auto size() const -> std::size_t {
+    // Count set bits in both enabled and disabled
+    std::size_t count = 0;
+    uint32_t combined = enabled_known | disabled_known;
+    while (combined) {
+      count += combined & 1;
+      combined >>= 1;
+    }
+    return count + custom.size();
+  }
+
+  /// Access vocabulary status by URI (throws if not found)
+  /// @param uri The vocabulary URI to look up
+  /// @return The boolean status of the vocabulary
+  /// @throws std::out_of_range if vocabulary not found
+  [[nodiscard]] auto at(std::string_view uri) const -> bool {
+    const auto status = find(uri);
+    if (!status.has_value()) {
+      throw std::out_of_range("Vocabulary not found: " + std::string(uri));
+    }
+    return status.value();
+  }
+
+  /// Check if there are no vocabularies
+  /// @return true if no vocabularies are set
+  [[nodiscard]] auto empty() const -> bool {
+    return enabled_known == 0 && disabled_known == 0 && custom.empty();
```

Comment:
Can you check if this can be safely `noexcept`? I think so, but I don't recall if `std::unordered_map::empty` can throw